### PR TITLE
Add typescript

### PR DESCRIPTION
**Pull request by AngyDev**
Related issue: #65

- I added the typescript configuration and the typescript lint checks in the package.json script.
- I added the index.d.ts file, but I'm not sure about some points. 

For example, the Constants type, the HalfEdgeHelper class, I added the code but it is commented because I'm not sure which is the correct type of the halfEdge parameter.

Maybe it's all a mess 😅, please tell me what you think.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    "noEmit": true,
+    "noUnusedParameters": true,
+    "strictNullChecks":true,
+	},
+  "include": [
+    "src"
+  ],
+  "exclude": [
+    "node_modules"
+  ]
```
Comment (gkjohnson) [inline tsconfig.json:21]: Can we make a the indentations consistent here?

File: .eslintrc.json [hunk 2]
```diff
@@ -1,4 +1,17 @@
...
+			"parser": "@typescript-eslint/parser",
+			"plugins": [
+				"@typescript-eslint"
+			],
+			"rules": {
+				"no-unused-vars": ["error", { "args": "none" }],
+				"indent": [2, 2]
+			}
+		}
+	]
```
Comment (gkjohnson) [inline .eslintrc.json:16]: More indentation fixes to make. I use `editorconfig` with vs code to automatically configure indentation.

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  updateFrom( geomtry: BufferGeometry ): void;
+
+}
+
+export class GridMaterial extends MeshPhongMaterial {
+
+  enableGrid: boolean;
+
+  onBeforeCompile( shader: ShaderMaterial ): void;
+  customProgramCacheKey(): boolean;
```
Comment (gkjohnson) [inline src/index.d.ts:110]: These two are defined on `MeshPhongMaterial` so there's no need to declare them on the types here.

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  setPoints( points: Vector3 ): void;
+
+}
+
+// export class HalfEdgeHelper extends EdgesHelper {
+
+//   constructor(geometry?: BufferGeometry, halfEdges?: );
+//   setHalfEdges(geometry: BufferGeometry, halfEdges: ): void;
+
+// }
```
Comment (gkjohnson) [inline src/index.d.ts:180]: Why is this one commented out?

**Review by gkjohnson — COMMENTED** [review]
This looks great! Thanks for adding it. Just a few comments to take a look at.

File: src/index.d.ts (hunk 4 above)
Comment (AngyDev) [reply src/index.d.ts:180]: Because I'm not sure which is the correct type for halfEdges param, THREE.HalfEdge?

File: tsconfig.json (hunk 1 above)
Comment (AngyDev) [reply tsconfig.json:21]: Sure, I don't know why I see correct in my editor and here it is not indented. I try to understand it.

File: .eslintrc.json (hunk 2 above)
Comment (AngyDev) [reply .eslintrc.json:16]: Thanks, I'll check

File: tsconfig.json (hunk 1 above)
Comment (gkjohnson) [reply tsconfig.json:21]: There's a mix of tab and space indentation here. Your editor might be configured so the width of the tabs makes the file look aligned. But Github uses different tab widths so it shows up differently.

File: src/index.d.ts (hunk 4 above)
Comment (gkjohnson) [reply src/index.d.ts:180]: This half edge map build internally is the one to pass in:

https://github.com/gkjohnson/three-bvh-csg/blob/main/src/core/HalfEdgeMap.js

**Comment by AngyDev** [comment]
The indentation issue was only about my formatter 😅
I did the changes that you suggested.
I hope it is correct

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  getGroupSet( index?: number ): Object;
+  getGroupArray( name: string, index?: number ): Array; // return Array ???
+  initializeArray( name: string, type: string ): void;
+  clear(): void;
+  delete( key: string ): void;
+  reset(): void;
+
+}
+
+export enum OperationStrategy{} // Is the name correct?
```
Comment (gkjohnson) [inline src/index.d.ts:29]: Why don't we call this "CSGOperation"

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  triangleSplitter: TriangleSplitter;
+  attributeData: TypedAttributeData;
+  attributes: String[];
+  useGroups: boolean;
+  debug: OperationDebugData;
+
+  evaluate( a: Brush, b: Brush, operation: OperationStrategy, targetBrush?: Brush ): Brush;
+
+  evaluateHierarchy( root: Brush, target?: Brush ): Brush; // Root is Brush type??
```
Comment (gkjohnson) [inline src/index.d.ts:45]: "root" should be "Operation" or "OperationGroup" here - heh this is good you're getting to things I haven't managed to document, yet 😁

Comment (AngyDev) [reply src/index.d.ts:45]: Ups, sorry 🤭!! I tried to understand the code and guess which are the correct types. 
If you think it is better I remove them.

Comment (gkjohnson) [reply src/index.d.ts]: No it's great! Thanks for including so much!

**Comment by gkjohnson** [comment]
Excellent, thank you again! This is a great addition.

If you're still interested in digging into the code a bit more adding some documentation for new classes #72 would be another good place to dive in - I can help provide insight on what some of these different classes do and why, as well. Of course now pressure, though, I appreciate the help!

**Pull request merged**